src/iso9660.rs

```diff
@@ -0,0 +1,364 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Minimal ISO 9660 parser. This is not a comprehensive parser; we only
+//! parse out the fields we care about. Extensions such as Rock Ridge
+//! and Joliet are not supported.
+//!
+//! The official specification is not free. The primary reference used
+//! for this module is https://wiki.osdev.org/ISO_9660.
+
+// An initial version of this module used the zerocopy crate to try to deserialize directly from
+// the on-disk ISO file in with zero copying. It works, but it's non-trivial and the performance
+// difference from just copying stuff didn't justify it.
+
+// Many magic numbers corresponding to offsets and lengths have not been const-ified. It should be
+// straightforward to see to what they correspond using the referenced linked above.
+
+use std::fs;
+use std::io::{Read, Seek, SeekFrom};
+use std::path::Path;
+
+use anyhow::{anyhow, bail, Context, Result};
+use bytes::{Buf, Bytes};
+use serde::Serialize;
+
+use crate::io::LimitReader;
+
+// technically the standard supports others, but this is the only one we support
+const ISO9660_SECTOR_SIZE: usize = 2048;
+
+#[derive(Debug, Serialize)]
+pub struct IsoFs {
+    descriptors: Vec<VolumeDescriptor>,
+    #[serde(skip_serializing)]
+    file: fs::File,
+}
+
+impl IsoFs {
+    pub fn new(path: impl AsRef<Path>) -> Result<Self> {
+        let mut file = fs::OpenOptions::new()
+            .read(true)
+            .open(path.as_ref())
+            .with_context(|| format!("opening {}", path.as_ref().display()))?;
+        let descriptors = get_volume_descriptors(&mut file)?;
+        Ok(Self { descriptors, file })
+    }
+
+    pub fn as_file(&mut self) -> Result<&mut fs::File> {
+        seek_to_sector(&mut self.file, 0).context("seeking to start of ISO")?;
+        Ok(&mut self.file)
+    }
+
+    pub fn get_root_directory(&self) -> Result<Directory> {
+        let primary = self
+            .get_primary_volume_descriptor()
+            .context("getting root directory")?;
+        Ok(primary.root.clone())
+    }
+
+    /// Returns an iterator over the records of a directory.
+    pub fn list_dir(&mut self, dir: &Directory) -> Result<IsoFsIterator> {
+        seek_to_sector(&mut self.file, dir.address)
+            .with_context(|| format!("seeking to directory {}", dir.name))?;
+
+        let mut buf = vec![0; dir.length as usize];
+        self.file
+            .read_exact(&mut buf)
+            .with_context(|| format!("reading directory {}", dir.name))?;
+
+        Ok(IsoFsIterator {
+            dir: Bytes::from(buf),
+        })
+    }
+
+    /// Returns the record for a specific path.
+    pub fn get_path(&mut self, path: &str) -> Result<DirectoryRecord> {
```

Comment:
Would it make sense to have a `get_file` helper method that rejects directories?